### PR TITLE
Publish snapshot packages to GitHub

### PR DESCRIPTION
**Pull request by porunov**
Discussion: https://lists.lfaidata.foundation/g/janusgraph-dev/topic/discussion_release_snapshot/84922650

Signed-off-by: Oleksandr Porunov <alexandr.porunov@gmail.com>

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there an issue associated with this PR? Is it referenced in the commit message?
- [ ] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### CODE REVIEW

File: docs/advanced-topics/snapshot-releases.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+snapshot releases and official releases.
+
+## Maven repository artifacts
+
+Official release are deployed to Sonatype OSS and are available in Sonatype Maven Central Repository. 
+No authentication is needed to use such releases.  
+
+Snapshot releases are deployed to GitHub Packages and are available in 
+[JanusGraph GitHub Packages registry](https://github.com/orgs/janusgraph/packages?repo_name=janusgraph). 
+Thus, such release are available only via GitHub authentication.
```
Comment (FlorianHockmann) [inline docs/advanced-topics/snapshot-releases.md:20]:
```suggestion
Thus, such releases are available only via GitHub authentication.  
```

File: docs/advanced-topics/snapshot-releases.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+Official release are deployed to Sonatype OSS and are available in Sonatype Maven Central Repository. 
+No authentication is needed to use such releases.  
+
+Snapshot releases are deployed to GitHub Packages and are available in 
+[JanusGraph GitHub Packages registry](https://github.com/orgs/janusgraph/packages?repo_name=janusgraph). 
+Thus, such release are available only via GitHub authentication.  
+
+To use snapshot releases in your Maven or Gradle build automation tool the next steps are required:
+1. Generate GitHub token with read access to packages and public repository access.
+2. Add JanusGraph GitHub Packages repository with higher priority to Maven Central repository.
```
Comment (FlorianHockmann) [inline docs/advanced-topics/snapshot-releases.md:24]:
```suggestion
2. Add JanusGraph GitHub Packages repository with higher priority than Maven Central repository. 
```
(?)

File: docs/advanced-topics/snapshot-releases.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+in case you don't want your token to be expired in the future.
+
+To create a token you can click on your profile -> Settings -> Developer settings -> Personal access tokens -> 
+Tokens (classic) -> Generate new token (classic).
+
+![](github_read_packages_token.png)
+
+After the token is generated it can be used to install JanusGraph public snapshot packages.
+
+#### Placing JanusGraph GitHub Packages repository
```
Comment (FlorianHockmann) [inline docs/advanced-topics/snapshot-releases.md:42]:
```suggestion
#### Adding JanusGraph GitHub Packages repository
```

(?)

File: docs/advanced-topics/snapshot-releases.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+Please, see detailed instruction on how to [define GitHub Package registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-apache-maven-registry#authenticating-to-github-packages).
+
+JanusGraph GitHub Packages repository after Maven Central Repository to have a higher priority. Otherwise the official
```
Comment (FlorianHockmann) [inline docs/advanced-topics/snapshot-releases.md:46]:
```suggestion
[Define/Place] JanusGraph GitHub Packages repository after Maven Central Repository to have a higher priority. Otherwise the official 
```

File: docs/advanced-topics/snapshot-releases.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+![](github_read_packages_token.png)
+
+After the token is generated it can be used to install JanusGraph public snapshot packages.
+
+#### Placing JanusGraph GitHub Packages repository
+
+Please, see detailed instruction on how to [define GitHub Package registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-apache-maven-registry#authenticating-to-github-packages).
+
+JanusGraph GitHub Packages repository after Maven Central Repository to have a higher priority. Otherwise the official 
+JanusGraph releases repository will be used which doesn't include any snapshot releases.
```
Comment (FlorianHockmann) [inline docs/advanced-topics/snapshot-releases.md:47]: So, maven doesn't go through all repositories configured until it finds the package with the specified version? It will instead only search the first repository where it finds the package and then fail if that repository doesn't have the version?

File: docs/advanced-topics/snapshot-releases.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+## JanusGraph distribution builds
+
+In addition to distribution builds provided for each official JanusGraph release, snapshot distribution builds are 
+provided for all commits.
+
+!!! info
+    GitHub allows to download distribution builds only for authenticated GitHub users.
+
+To access distribution build bundle for any commit, please, open commit you are interested in and select it's
```
Comment (FlorianHockmann) [inline docs/advanced-topics/snapshot-releases.md:116]:
```suggestion
To access the distribution build bundle for any commit, please open the commit you are interested in and select its 
```

File: .github/workflows/ci-snapshot-publish.yml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+      - uses: actions/setup-java@v3
+        with:
+          java-version: 8
+          distribution: zulu
+
+      - name: Setup unique version
+        run:  mvn versions:set -DnewVersion="$(mvn help:evaluate -Dexpression=project.version -q -DforceStdout)-$(git rev-parse --short HEAD)" -DgenerateBackupPoms=false
+
+      - name: Compile and prepare snapshot release
+        run: mvn clean install -Pjanusgraph-release -Dgpg.skip=true -DskipTests=true
```
Comment (FlorianHockmann) [inline .github/workflows/ci-snapshot-publish.yml:45]: Wouldn't it be better to not skip GPG signing here (and in the deploy step below)? Or is there any reason why we cannot do that here?

**Review by FlorianHockmann — CHANGES_REQUESTED** [review]
Thanks for implementing this and especially also for writing the documentation!
I added some comments, but most are only about small grammar issues.

File: docs/advanced-topics/snapshot-releases.md (hunk 5 above)
Comment (porunov) [reply docs/advanced-topics/snapshot-releases.md:47]: That's right. At least that's what I experienced with Gradle. Could be that it's different in Maven. I will double check on Tuesday and correct documentation if it's different for Maven. 
Also, I don't know if there is a way telling Maven a fallback repository.
If we place JanusGraph before Maven Central in Gradle - it will use only Maven Central for JanusGraph. If we place it vice versa - it will use only JanusGraph GitHub repository for JanusGraph dependencies, and not Maven Central.

**Comment by porunov** [comment]
> Thanks for implementing this and especially also for writing the documentation!
> I added some comments, but most are only about small grammar issues.

Thank you! I will apply the changes on Monday or Tuesday.

**Comment by porunov** [comment]
@FlorianHockmann what do you think if we publish JanusGraph snapshot releases to Maven Central as well but we use different `groupId` to publish JanusGraph releases? I.e. I guess we could use something like `org.janusgraph.snapshot` instead of `org.janusgraph`.
In such case instead of creating a GitHub token and adding the JanusGraph GitHub repository to Maven / Gradle users will be able to just change `groupId` and `artifactId` for their dependencies.
For example, their dependency could look like the next one:
```
<dependency>
    <groupId>org.janusgraph.snapshot</groupId>
    <artifactId>janusgraph-core</artifactId>
    <version>0.6.3-SNAPSHOT-39839b810</version>
</dependency>
```
For some reason I thought that releasing too many versions to Sonatype could be bad, but looking at some example (Like: [aws s3](https://mvnrepository.com/artifact/com.amazonaws/aws-java-sdk-s3)) - it looks like it's fine to make releases to Sonatype every day or multiple times a day.
The only thing is that we will need to create a new groupId in sonatype, give permissions to current release managers, and correctly setup release process (i.e. to automatically close and release artifacts instead of doing the manual work).
Potentially this could be done via `nexus-staging-maven-plugin`: https://help.sonatype.com/repomanager2/staging-releases/configuring-your-project-for-deployment

Basically, it feels that releasing to GitHub is easier on our side to be implemented but harder for users to use those releases. With Maven Central it feels vice versa. Slightly harder for us to be implemented but easier for users to use those releases.

**Comment by porunov** [comment]
@FlorianHockmann it was easier than I thought. I created a replacement PR to this PR here: #3364 
I applied all your comments in that PR. Description is now simplified. With that PR users will be able to use JanusGraph snapshot releases from Maven Central. Thus, no authentication is needed.
